### PR TITLE
esy: init at 0.6.6

### PR DESCRIPTION
**Pull request by xavierzwirtz**
###### Motivation for this change

Packaging of the esy package manager, by building on #65098, #85835 and esy/esy#994.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by xavierzwirtz** [comment]
cc @NickHu, @lobre

**Comment by NickHu** [comment]
I'm not in favor of using an FHS root for this, but I'm also not using reason anymore so feel free to discard my opinion.

**Comment by xavierzwirtz** [comment]
Benefit of the FHS root is the OPAM ecosystem will work without requiring changes to any packages. I do not have an amazing amount of experience with Nix though, so I am unsure what the trade-offs are with this solution. I copied this approach from the flutter compiler.

File: pkgs/development/ocaml-modules/cudf/default.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+  buildPhase = ''
+    make all opt
+  '';
+
+  createFindlibDestdir = true;
+
+  meta = {
+    description = "CUDF (for Common Upgradeability Description Format) is a format for describing upgrade scenarios in package-based Free and Open Source Software distribution.";
+    license = stdenv.lib.licenses.lgpl3;
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/cudf/default.nix:39]: Please add yourself as a maintainer.

File: pkgs/development/ocaml-modules/cudf/default.nix [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  ];
+
+  buildPhase = ''
+    make all opt
+  '';
+
+  createFindlibDestdir = true;
+
+  meta = {
+    description = "CUDF (for Common Upgradeability Description Format) is a format for describing upgrade scenarios in package-based Free and Open Source Software distribution.";
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/cudf/default.nix:38]:
```suggestion
    description = "Format for describing upgrade scenarios in package-based Free and Open Source Software distribution";
```

File: pkgs/development/ocaml-modules/cudf/default.nix (hunk 1 above)
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/cudf/default.nix:39]:
```suggestion
    license = lib.licenses.lgpl3;
```

plus or only?

File: pkgs/development/ocaml-modules/dose3/default.nix [hunk 3]
```diff
@@ -0,0 +1,66 @@
+{ stdenv, fetchurl, ocaml, ocamlbuild, findlib
+, cppo, cudf, ocamlgraph, ocaml_extlib, re, perl, ncurses }:
+
+let base_patch_url = "https://raw.githubusercontent.com/ocaml/opam-repository/master/packages/dose3/dose3.5.0.1/files";
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/dose3/default.nix:4]: Please replace the version number with the version variable.

File: pkgs/development/ocaml-modules/dose3/default.nix [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  preInstall = ''
+    substituteInPlace Makefile.config \
+      --replace "-destdir \$(LIBDIR)" ""
+  '';
+
+  createFindlibDestdir = true;
+
+  meta = {
+    description = "Dose library (part of Mancoosi tools).";
+    license = stdenv.lib.licenses.lgpl3;
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/dose3/default.nix:63]: Please add yourself as a maintainer.

```suggestion
    license = lib.licenses.lgpl3;
```

plus or only?

File: pkgs/development/tools/ocaml/esy/solve-cudf.nix [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  postInstall = ''
+    mkdir $npm
+    set -- $srcs
+    cp $2 $npm/package.json
+  '';
+
+  meta = {
+    homepage = https://github.com/andreypopp/esy-solve-cudf;
+    description = "package.json workflow for native development with Reason/OCaml";
+    license = stdenv.lib.licenses.gpl3;
```
Comment (SuperSandro2000) [inline pkgs/development/tools/ocaml/esy/solve-cudf.nix:50]: same as above.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**